src/grpc/infer_handler.h

```diff
@@ -845,33 +846,37 @@ class InferHandlerState {
             << " step " << state->step_;
         return true;
       }
-      if ((state->step_ != Steps::CANCELLATION_ISSUED) &&
-          (state->step_ != Steps::CANCELLED)) {
+      if (is_notification) {
         LOG_VERBOSE(1) << "Cancellation notification received for " << name
                        << ", rpc_ok=" << rpc_ok << ", context "
                        << state->context_->unique_id_ << ", "
                        << state->unique_id_ << " step " << state->step_;
+      }
 
-        // If the context has not been cancelled then
-        // issue cancellation request to all the inflight
-        // states belonging to the context.
-        if (state->context_->step_ != Steps::CANCELLED) {
-          IssueRequestCancellation();
-          // Mark the context as cancelled
-          state->context_->step_ = Steps::CANCELLED;
-          // The state returns true because the CancelExecution
-          // call above would have raised alarm objects on all
-          // pending inflight states objects. This state will
-          // be taken up along with all the other states in the
-          // next iteration from the completion queue which
-          // would release the state.
-          return true;
-        }
+      // If the context has not been cancelled then
+      // issue cancellation request to all the inflight
+      // states belonging to the context.
+      // It means this is the first time we are hiting this line for this grpc
+      // transaction.
+      if ((state->step_ != Steps::CANCELLATION_ISSUED) &&
+          (state->step_ != Steps::CANCELLED) &&
+          (state->context_->step_ != Steps::CANCELLED)) {
+        IssueRequestCancellation();
+        // Mark the context as cancelled
+        state->context_->step_ = Steps::CANCELLED;
+        // The state returns true because the CancelExecution
+        // call above would have raised alarm objects on all
+        // pending inflight states objects. This state will
+        // be taken up along with all the other states in the
+        // next iteration from the completion queue which
+        // would release the state.
+        return true;
       }
 
-      if (state->step_ != Steps::CANCELLATION_ISSUED) {
-        // The cancellation has not been issued hence the state can
-        // be released.
+      if (state->step_ != Steps::CANCELLATION_ISSUED &&
```

Comment:
For ease of readability. can you re-organize the logic in L861 as:

```
if (state->step_ != Steps::CANCELLATION_ISSUED) {
     if ((state->step_ != Steps::CANCELLED) && (state->context_->step_ != Steps::CANCELLED)) {
         // Issue the request cancellation as it has not been cancelled yet.
          ....
     } else if (!(is_notification && state->step_ == Steps::CANCELLED)) {
         // @yinggeh  describe when we will land in this block.
          .....
     }
} else { // state->step_ == Steps::CANCELLATION_ISSUED
 // Should wait for the ResponseComplete callbacks to be invoked
}
```

Reply:
Updated